File: source/modulo_components/test/cpp/test_component_interface_empty_parameters.cpp
#include <gtest/gtest.h>

#include "modulo_components/exceptions/ComponentParameterException.hpp"
#include "modulo_core/EncodedState.hpp"
#include "test_modulo_components/component_public_interfaces.hpp"

namespace modulo_components {

template<class NodeT>
class EmptyParameterInterface : public ComponentInterfacePublicInterface<NodeT> {
public:
  explicit EmptyParameterInterface(
      const rclcpp::NodeOptions& node_options, modulo_core::communication::PublisherType publisher_type,
      const std::string& fallback_name = "EmptyParameterInterface", bool allow_empty = true, bool add_parameter = true
  ) : ComponentInterfacePublicInterface<NodeT>(node_options, publisher_type, fallback_name), allow_empty_(allow_empty) {
    if (add_parameter) {
      this->add_parameter(std::make_shared<Parameter<std::string>>("name"), "Test parameter");
    }
  };

private:
  bool validate_parameter(const std::shared_ptr<state_representation::ParameterInterface>& parameter) override {
    if (parameter->get_name() == "name") {
      if (parameter->is_empty()) {
        return this->allow_empty_;
      } else if (parameter->get_parameter_value<std::string>().empty()) {
        RCLCPP_ERROR(this->get_logger(), "Provide a non empty value for parameter 'name'");
        return false;
      }
    }
    return true;
  }

  bool allow_empty_;
};

template<class NodeT>
class ComponentInterfaceEmptyParameterTest : public ::testing::Test {
protected:
  static void SetUpTestSuite() {
    rclcpp::init(0, nullptr);
  }

  static void TearDownTestSuite() {
    rclcpp::shutdown();
  }

  void SetUp() override {
    if (std::is_same<NodeT, rclcpp::Node>::value) {
      this->component_ = std::make_shared<EmptyParameterInterface<NodeT>>(
          rclcpp::NodeOptions(), modulo_core::communication::PublisherType::PUBLISHER
      );
    } else if (std::is_same<NodeT, rclcpp_lifecycle::LifecycleNode>::value) {
      this->component_ = std::make_shared<EmptyParameterInterface<NodeT>>(
          rclcpp::NodeOptions(), modulo_core::communication::PublisherType::LIFECYCLE_PUBLISHER
      );
    }
  }

  std::shared_ptr<EmptyParameterInterface<NodeT>> component_;
};
using NodeTypes = ::testing::Types<rclcpp::Node, rclcpp_lifecycle::LifecycleNode>;
TYPED_TEST_SUITE(ComponentInterfaceEmptyParameterTest, NodeTypes);

TYPED_TEST(ComponentInterfaceEmptyParameterTest, NotAllowEmptyOnConstruction) {
  if (std::is_same<TypeParam, rclcpp::Node>::value) {
    EXPECT_THROW(std::make_shared<EmptyParameterInterface<TypeParam>>(
        rclcpp::NodeOptions(), modulo_core::communication::PublisherType::PUBLISHER, "EmptyParameterComponent", false
    ), modulo_components::exceptions::ComponentParameterException);
  } else if (std::is_same<TypeParam, rclcpp_lifecycle::LifecycleNode>::value) {
    EXPECT_THROW(std::make_shared<EmptyParameterInterface<TypeParam>>(
        rclcpp::NodeOptions(), modulo_core::communication::PublisherType::LIFECYCLE_PUBLISHER,
        "EmptyParameterComponent", false
    ), modulo_components::exceptions::ComponentParameterException);
  }
}

TYPED_TEST(ComponentInterfaceEmptyParameterTest, NotAllowEmpty) {
  std::shared_ptr<EmptyParameterInterface<TypeParam>> component;
  if (std::is_same<TypeParam, rclcpp::Node>::value) {
    component = std::make_shared<EmptyParameterInterface<TypeParam>>(
        rclcpp::NodeOptions(), modulo_core::communication::PublisherType::PUBLISHER, "EmptyParameterComponent", false,
        false
    );
  } else if (std::is_same<TypeParam, rclcpp_lifecycle::LifecycleNode>::value) {
    component = std::make_shared<EmptyParameterInterface<TypeParam>>(
        rclcpp::NodeOptions(), modulo_core::communication::PublisherType::LIFECYCLE_PUBLISHER,
        "EmptyParameterComponent", false, false
    );
  }
  EXPECT_THROW(component->add_parameter(std::make_shared<Parameter<std::string>>("name"), "Test parameter"),
               exceptions::ComponentParameterException);
  EXPECT_THROW(auto param = component->get_parameter("name"), exceptions::ComponentParameterException);
  EXPECT_THROW(component->get_ros_parameter("name"), rclcpp::exceptions::ParameterNotDeclaredException);
}

TYPED_TEST(ComponentInterfaceEmptyParameterTest, AllowEmpty) {
  std::shared_ptr<EmptyParameterInterface<TypeParam>> component;
  if (std::is_same<TypeParam, rclcpp::Node>::value) {
    component = std::make_shared<EmptyParameterInterface<TypeParam>>(
        rclcpp::NodeOptions(), modulo_core::communication::PublisherType::PUBLISHER, "EmptyParameterComponent", true,
        false
    );
  } else if (std::is_same<TypeParam, rclcpp_lifecycle::LifecycleNode>::value) {
    component = std::make_shared<EmptyParameterInterface<TypeParam>>(
        rclcpp::NodeOptions(), modulo_core::communication::PublisherType::LIFECYCLE_PUBLISHER,
        "EmptyParameterComponent", true, false
    );
  }
  EXPECT_NO_THROW(component->add_parameter(std::make_shared<Parameter<std::string>>("name"), "Test parameter"));
  EXPECT_EQ(component->get_parameter("name")->get_parameter_type(), ParameterType::STRING);
  EXPECT_TRUE(component->get_parameter("name")->is_empty());
  EXPECT_EQ(component->get_ros_parameter("name").get_type(), rclcpp::PARAMETER_NOT_SET);
}

TYPED_TEST(ComponentInterfaceEmptyParameterTest, ValidateEmptyParameter) {
  // component comes with empty parameter 'name'
  EXPECT_EQ(this->component_->get_parameter("name")->get_parameter_type(), ParameterType::STRING);
  EXPECT_TRUE(this->component_->get_parameter("name")->is_empty());
  auto ros_param = rclcpp::Parameter();
  EXPECT_NO_THROW(ros_param = this->component_->get_ros_parameter("name"););
  EXPECT_EQ(ros_param.get_type(), rclcpp::ParameterType::PARAMETER_NOT_SET);

  // Trying to overwrite a parameter is not possible
  this->component_->add_parameter(std::make_shared<Parameter<bool>>("name"), "Test parameter");
  EXPECT_EQ(this->component_->get_parameter("name")->get_parameter_type(), ParameterType::STRING);
  EXPECT_TRUE(this->component_->get_parameter("name")->is_empty());
  EXPECT_NO_THROW(ros_param = this->component_->get_ros_parameter("name"););
  EXPECT_EQ(ros_param.get_type(), rclcpp::ParameterType::PARAMETER_NOT_SET);

  // Set parameter value from ROS interface should update ROS parameter type
  this->component_->set_ros_parameter({"name", "test"});
  EXPECT_EQ(this->component_->get_parameter("name")->get_parameter_type(), ParameterType::STRING);
  EXPECT_EQ(this->component_->get_parameter("name")->template get_parameter_value<std::string>(), "test");
  EXPECT_NO_THROW(ros_param = this->component_->get_ros_parameter("name"););
  EXPECT_EQ(ros_param.get_type(), rclcpp::ParameterType::PARAMETER_STRING);
  EXPECT_EQ(ros_param.template get_value<std::string>(), "test");

  // Set parameter value from component interface
  this->component_->template set_parameter_value<std::string>("name", "again");
  EXPECT_EQ(this->component_->get_parameter("name")->get_parameter_type(), ParameterType::STRING);
  EXPECT_EQ(this->component_->get_parameter("name")->template get_parameter_value<std::string>(), "again");
  EXPECT_NO_THROW(ros_param = this->component_->get_ros_parameter("name"););
  EXPECT_EQ(ros_param.get_type(), rclcpp::ParameterType::PARAMETER_STRING);
  EXPECT_EQ(ros_param.template get_value<std::string>(), "again");

  // Setting it with empty value should be rejected in parameter evaluation
  this->component_->template set_parameter_value<std::string>("name", "");
  EXPECT_EQ(this->component_->get_parameter("name")->get_parameter_type(), ParameterType::STRING);
  EXPECT_EQ(this->component_->get_parameter("name")->template get_parameter_value<std::string>(), "again");
  EXPECT_NO_THROW(ros_param = this->component_->get_ros_parameter("name"););
  EXPECT_EQ(ros_param.get_type(), rclcpp::ParameterType::PARAMETER_STRING);
  EXPECT_EQ(ros_param.template get_value<std::string>(), "again");

  // Setting it with empty value should be rejected in parameter evaluation
  this->component_->set_ros_parameter({"name", ""});
  EXPECT_EQ(this->component_->get_parameter("name")->get_parameter_type(), ParameterType::STRING);
  EXPECT_EQ(this->component_->get_parameter("name")->template get_parameter_value<std::string>(), "again");
  EXPECT_NO_THROW(ros_param = this->component_->get_ros_parameter("name"););
  EXPECT_EQ(ros_param.get_type(), rclcpp::ParameterType::PARAMETER_STRING);
  EXPECT_EQ(ros_param.template get_value<std::string>(), "again");

  // TODO clarify that behavior somewhere
  // Setting a parameter with type NOT_SET undeclares that parameter
  this->component_->set_ros_parameter(rclcpp::Parameter("name", rclcpp::ParameterValue{}));
  EXPECT_THROW(this->component_->get_ros_parameter("name"), rclcpp::exceptions::ParameterNotDeclaredException);
}

TYPED_TEST(ComponentInterfaceEmptyParameterTest, ChangeParameterType) {
  // Add parameter from component interface
  this->component_->add_parameter(std::make_shared<Parameter<int>>("int"), "Test parameter");
  EXPECT_TRUE(this->component_->describe_parameter("int").dynamic_typing);
  this->component_->template set_parameter_value<int>("int", 1);
  EXPECT_EQ(this->component_->get_parameter("int")->get_parameter_type(), ParameterType::INT);
  EXPECT_EQ(this->component_->get_parameter("int")->template get_parameter_value<int>(), 1);
  auto ros_param = rclcpp::Parameter();
  EXPECT_NO_THROW(ros_param = this->component_->get_ros_parameter("int"););
  EXPECT_EQ(ros_param.get_type(), rclcpp::ParameterType::PARAMETER_INTEGER);
  EXPECT_EQ(ros_param.template get_value<int>(), 1);

  // Set parameter value from component interface with different type should not work
  this->component_->template set_parameter_value<double>("int", 2.0);
  EXPECT_EQ(this->component_->get_parameter("int")->get_parameter_type(), ParameterType::INT);
  EXPECT_EQ(this->component_->get_parameter("int")->template get_parameter_value<int>(), 1);
  EXPECT_NO_THROW(ros_param = this->component_->get_ros_parameter("int"););
  EXPECT_EQ(ros_param.get_type(), rclcpp::ParameterType::PARAMETER_INTEGER);
  EXPECT_EQ(ros_param.template get_value<int>(), 1);

  // Set parameter value from ROS interface with different type should not work
  auto result = this->component_->set_ros_parameter(rclcpp::Parameter("int", 2.0));
  EXPECT_FALSE(result.successful);
  EXPECT_EQ(this->component_->get_parameter("int")->get_parameter_type(), ParameterType::INT);
  EXPECT_EQ(this->component_->get_parameter("int")->template get_parameter_value<int>(), 1);
  EXPECT_NO_THROW(ros_param = this->component_->get_ros_parameter("int"););
  EXPECT_EQ(ros_param.get_type(), rclcpp::ParameterType::PARAMETER_INTEGER);
  EXPECT_EQ(ros_param.template get_value<int>(), 1);
}

TYPED_TEST(ComponentInterfaceEmptyParameterTest, ParameterOverrides) {
  // Construction with allowing empty parameters but providing the parameter override should succeed
  if (std::is_same<TypeParam, rclcpp::Node>::value) {
    EXPECT_NO_THROW(std::make_shared<EmptyParameterInterface<TypeParam>>(
        rclcpp::NodeOptions().parameter_overrides({rclcpp::Parameter("name", "test")}),
        modulo_core::communication::PublisherType::PUBLISHER, "EmptyParameterComponent", false
    ));
  } else if (std::is_same<TypeParam, rclcpp_lifecycle::LifecycleNode>::value) {
    EXPECT_NO_THROW(std::make_shared<EmptyParameterInterface<TypeParam>>(
        rclcpp::NodeOptions().parameter_overrides({rclcpp::Parameter("name", "test")}),
        modulo_core::communication::PublisherType::LIFECYCLE_PUBLISHER, "EmptyParameterComponent", false
    ));
  }
}
} // namespace modulo_components

File: source/modulo_components/test/python/test_component_interface_empty_parameters.py
import pytest
import rclpy
import state_representation as sr
from modulo_components.component_interface import ComponentInterface
from modulo_components.exceptions import ComponentParameterError
from rcl_interfaces.msg import SetParametersResult
from rclpy.exceptions import ParameterNotDeclaredException
from rclpy.node import Node
from rclpy.parameter import Parameter


class EmtpyParameterInterface(ComponentInterface):
    def __init__(self, node_name, allow_empty=True, add_parameter=True, empty_parameter=True, *kargs, **kwargs):
        super().__init__(node_name, *kargs, **kwargs)
        self._allow_empty = allow_empty
        if add_parameter:
            if empty_parameter:
                self.add_parameter(sr.Parameter("name", sr.ParameterType.STRING), "Test parameter")
            else:
                self.add_parameter(sr.Parameter("name", "test", sr.ParameterType.STRING), "Test parameter")

    def get_ros_parameter(self, name: str) -> rclpy.Parameter:
        return rclpy.node.Node.get_parameter(self, name)

    def set_ros_parameter(self, param: rclpy.Parameter) -> SetParametersResult:
        return rclpy.node.Node.set_parameters(self, [param])[0]

    def on_validate_parameter_callback(self, parameter: sr.Parameter) -> bool:
        if parameter.get_name() == "name":
            if parameter.is_empty():
                return self._allow_empty
            elif not parameter.get_value():
                self.get_logger().error("Provide a non empty value for parameter 'name'")
                return False
        return True


@pytest.fixture()
def component_test(ros_context):
    yield EmtpyParameterInterface("component")


def test_not_allow_empty_on_construction(ros_context):
    # Construction with empty parameter should raise if the empty parameter is not allowed
    with pytest.raises(ComponentParameterError):
        EmtpyParameterInterface("component", False)


def test_not_allow_empty(ros_context):
    component = EmtpyParameterInterface("component", allow_empty=False, add_parameter=False)
    with pytest.raises(ComponentParameterError):
        component.add_parameter(sr.Parameter("name", sr.ParameterType.STRING), "Test parameter")
    with pytest.raises(ComponentParameterError):
        component.get_parameter("name")
    with pytest.raises(ParameterNotDeclaredException):
        assert Node.get_parameter(component, "name")

    component = EmtpyParameterInterface("component", allow_empty=True, add_parameter=False)
    component.add_parameter(sr.Parameter("name", sr.ParameterType.STRING), "Test parameter")
    assert component.get_parameter("name").get_parameter_type() == sr.ParameterType.STRING
    assert component.get_parameter("name").is_empty()
    assert Node.get_parameter(component, "name").type_ == Parameter.Type.NOT_SET


def test_validate_empty_parameter(component_test):
    # component_test comes with empty parameter 'name'
    assert component_test.get_parameter("name").get_parameter_type() == sr.ParameterType.STRING
    assert component_test.get_parameter("name").is_empty()
    assert Node.get_parameter(component_test, "name").type_ == Parameter.Type.NOT_SET

    # Trying to overwrite a parameter is not possible
    component_test.add_parameter(sr.Parameter("name", sr.ParameterType.BOOL), "Test parameter")
    assert component_test.get_parameter("name").get_parameter_type() == sr.ParameterType.STRING
    assert component_test.get_parameter("name").is_empty()
    assert Node.get_parameter(component_test, "name").type_ == Parameter.Type.NOT_SET

    # Set parameter with empty parameter isn't possible because there is no method for that
    # component_test.set_parameter(sr.Parameter("name", sr.ParameterType.STRING))

    # Set parameter value from ROS interface should update ROS parameter type
    Node.set_parameters(component_test, [Parameter("name", value="test")])
    assert component_test.get_parameter("name").get_parameter_type() == sr.ParameterType.STRING
    assert component_test.get_parameter_value("name") == "test"
    assert Node.get_parameter(component_test, "name").type_ == Parameter.Type.STRING
    assert Node.get_parameter(component_test, "name").value == "test"

    # Set parameter value from component interface
    component_test.set_parameter_value("name", "again", sr.ParameterType.STRING)
    assert component_test.get_parameter("name").get_parameter_type() == sr.ParameterType.STRING
    assert component_test.get_parameter_value("name") == "again"
    assert Node.get_parameter(component_test, "name").type_ == Parameter.Type.STRING
    assert Node.get_parameter(component_test, "name").value == "again"

    # Setting it with empty value should be rejected in parameter evaluation
    component_test.set_parameter_value("name", "", sr.ParameterType.STRING)
    assert component_test.get_parameter("name").get_parameter_type() == sr.ParameterType.STRING
    assert component_test.get_parameter_value("name") == "again"
    assert Node.get_parameter(component_test, "name").type_ == Parameter.Type.STRING
    assert Node.get_parameter(component_test, "name").value == "again"

    # Setting it with empty value should be rejected in parameter evaluation
    Node.set_parameters(component_test, [Parameter("name", type_=Parameter.Type.STRING, value="")])
    assert component_test.get_parameter("name").get_parameter_type() == sr.ParameterType.STRING
    assert component_test.get_parameter_value("name") == "again"
    assert Node.get_parameter(component_test, "name").type_ == Parameter.Type.STRING
    assert Node.get_parameter(component_test, "name").value == "again"

    # TODO clarify that behavior somewhere
    # Setting a parameter with type NOT_SET undeclares that parameter
    Node.set_parameters(component_test, [Parameter("name", type_=Parameter.Type.NOT_SET)])
    with pytest.raises(ParameterNotDeclaredException):
        Node.get_parameter(component_test, "name")


def test_change_parameter_type(component_test):
    # Add parameter from component interface
    component_test.add_parameter(sr.Parameter("int", sr.ParameterType.INT), "Test parameter")
    assert component_test.describe_parameter("int").dynamic_typing
    component_test.set_parameter_value("int", 1, sr.ParameterType.INT)
    assert component_test.get_parameter("int").get_parameter_type() == sr.ParameterType.INT
    assert component_test.get_parameter_value("int") == 1
    ros_param = component_test.get_ros_parameter("int")
    assert ros_param.type_ == Parameter.Type.INTEGER
    assert ros_param.value == 1

    # Set parameter value from component interface with different type should not work
    component_test.set_parameter_value("int", 2.0, sr.ParameterType.DOUBLE)
    assert component_test.get_parameter("int").get_parameter_type() == sr.ParameterType.INT
    assert component_test.get_parameter_value("int") == 1
    ros_param = component_test.get_ros_parameter("int")
    assert ros_param.type_ == Parameter.Type.INTEGER
    assert ros_param.value == 1

    # Set parameter value from ROS interface with different type should not work
    component_test.set_ros_parameter(Parameter("int", value=2.0))
    assert component_test.get_parameter("int").get_parameter_type() == sr.ParameterType.INT
    assert component_test.get_parameter_value("int") == 1
    ros_param = component_test.get_ros_parameter("int")
    assert ros_param.type_ == Parameter.Type.INTEGER
    assert ros_param.value == 1


def test_parameter_overrides(ros_context):
    # Construction with not allowing empty parameters but providing the parameter override should succeed
    EmtpyParameterInterface("component", False, parameter_overrides=[Parameter("name", value="test")])


def test_parameter_overrides_empty(ros_context):
    # Construction with not allowing empty parameters and providing an uninitialized parameter override should not succeed
    with pytest.raises(ComponentParameterError):
        EmtpyParameterInterface("component", allow_empty=False, empty_parameter=False,
                                parameter_overrides=[Parameter("name")])
